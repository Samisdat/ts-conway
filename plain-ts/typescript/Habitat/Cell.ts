import {CellType} from '@Conway/Frontend/CellType/CellType';
import {CellUnkown} from '@Conway/Frontend/CellType/CellUnkown';
import {CellTypesFactory} from '@Conway/Frontend/CellType/CellTypesFactory';
import {Position} from '@Conway/Geometry/Position';

export class Cell {

    public readonly relativePosition: Position;
    public readonly absolutePosition: Position;
    public readonly offset: Position;

    private type: CellType = new CellUnkown();

    constructor(
        relativePosition: Position,
        absolutePosition: Position,
        offset: Position
    ) {

        this.relativePosition = relativePosition;

        this.absolutePosition = absolutePosition;

        this.offset = offset;

        this.setCheckerboardColor();

    }

    private setCheckerboardColor(): void {

        const x = this.absolutePosition.x;
        const y = this.absolutePosition.y;

        this.setType(CellTypesFactory.get().checkerboardDark());

        if ( (0 === y % 2 && 0 === x % 2 ) || (0 !== y % 2 && 0 !== x % 2)) {
            this.setType(CellTypesFactory.get().checkerboardLight());
        }

        if (0 === x && 0 === y) {
            this.setType(CellTypesFactory.get().center());
        }

    }

    public setType(type: CellType): void {
        this.type = type;
    }

    public getType(): CellType {
        return this.type;
    }

    public getColor(): string {
        return this.type.hex;
    }

    get x(): number {
        return this.relativePosition.x + this.offset.x;
    }
    get y(): number {
        return this.relativePosition.y + this.offset.y;
    }

}
