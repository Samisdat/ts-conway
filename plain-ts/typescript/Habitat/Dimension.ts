export class Dimension {

    private readonly _rows: number;
    private readonly _cols: number;

    constructor(rows: number, cols: number) {

        this.ensurePositiveNumber(rows, 'rows');
        this.ensurePositiveNumber(cols, 'cols');

        this._rows = rows;
        this._cols = cols;

    }

    public getRows(): number {
        return this._rows;
    }

    public getCols(): number {
        return this._cols;
    }

    private ensurePositiveNumber(value: number, attributeName: string) {

        if (1 > value) {
            throw new Error(attributeName + ' must be at least 1');
        }


    }
}
