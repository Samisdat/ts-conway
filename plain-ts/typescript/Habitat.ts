import {Position} from '@Conway/Geometry/Position';
import {Population} from '@Conway/Population';
import {Cell} from '@Conway/Habitat/Cell';
import {CellTypesFactory} from '@Conway/Frontend/CellType/CellTypesFactory';
import {Dimension} from '@Conway/Habitat/Dimension';

interface MatrixCol {
    [index: number]: number;
}

interface MatrixRow {
    [index: number]: MatrixCol;
}

export class Habitat {

    private gridDimension: Dimension;
    private sourcePosition: Position;

    private offset: Position;

    private cells: Cell[] = [];

    public readonly center: Position;

    private absoluteMap: MatrixRow = {};

    constructor(
        population: Population,
        gridDimension: Dimension,
        sourcePosition: Position,
        offset: Position
    ) {

        Habitat.ensureOffsetBetweenMinusOneAndOne(offset);

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        let relativePointer = new Position(0, 0);
        let absolutePointer = this.sourcePosition as Position;

        const inMiddleOfGrid = new Position(
            Math.floor(this.gridDimension.getRows() / 2),
            Math.floor(this.gridDimension.getCols() / 2)
        );

        this.center = relativePointer.move(inMiddleOfGrid);

        relativePointer = relativePointer.move(inMiddleOfGrid.inverse());

        absolutePointer = absolutePointer.move(inMiddleOfGrid.inverse());

        this.createGridCells(relativePointer, absolutePointer);

        this.markLivingCells(population);

    }

    private createGridCells(relativePointer: Position, absolutePointer: Position){

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.getCols(); top += 1) {

            for (let right = 0; right < this.gridDimension.getRows(); right += 1) {

                const gridCell = new Cell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new Position(-1 * this.gridDimension.getRows(), 1),
            );

            absolutePointer = absolutePointer.move(
                new Position(-1 * this.gridDimension.getRows(), 1),
            );

        }

    }

    private markLivingCells(population: Population){

        for (const positionWithLivingCells of population.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private static ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): Position {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.getCols();
    }

    public getRows(): number {
        return this.gridDimension.getRows();
    }

    public getCells() {
        return this.cells;
    }

}
