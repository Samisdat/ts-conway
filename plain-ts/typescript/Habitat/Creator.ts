import {Position} from '@Conway/Geometry/Position';
import {Dimension} from '@Conway/Habitat/Dimension';

export class Creator {

    public readonly canvasWidth: number;
    public readonly canvasHeigth: number;
    public readonly originalCellWidthAndHeight: number;
    public readonly cellWidthAndHeight: number;
    public readonly pan: Position;

    public readonly sourcePosition: Position;
    public readonly offset: Position;

    public readonly zoom: number;

    private dimension: Dimension;

    constructor(
        canvasWidth: number,
        canvasHeight: number,
        cellWidthAndHeight: number,
        pan: Position,
        zoom: number
    ) {
        this.canvasWidth = canvasWidth;
        this.canvasHeigth = canvasHeight;
        this.originalCellWidthAndHeight = cellWidthAndHeight;
        this.pan = pan;
        this.zoom = zoom;

        this.offset = new Position(
            this.pan.x % 1,
            this.pan.y % 1,

        );

        const sourcePosition = this.pan.clone().move(this.offset.inverse());
        this.sourcePosition = new Position(
            sourcePosition.x,
            sourcePosition.y
        );

        this.cellWidthAndHeight = this.originalCellWidthAndHeight * this.getZoom();

        this.setDimension();

    }

    private setDimension(): void{

        let rows = Math.ceil(this.canvasWidth / this.cellWidthAndHeight);

        if (0 === rows % 2) {
            rows += 1;
        }

        if (0 !== this.offset.x) {
            rows += 2;
        }

        let cols = Math.ceil(this.canvasHeigth / this.cellWidthAndHeight);

        if (0 === cols % 2) {
            cols += 1;
        }

        if (0 !== this.offset.y) {
            cols += 2;
        }

        this.dimension = new Dimension(rows, cols);

    }

    public getRows(): number {
        return this.dimension.getRows();
    }

    public getCols(): number {
        return this.dimension.getCols();
    }

    public getPan(): Position {
        return this.pan;
    }

    public getSourcePosition(): Position {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset.inverse();
    }

    public getZoom(): number {
        return this.zoom;
    }

    public getDimension(): Dimension {
        return this.dimension;
    }

}
