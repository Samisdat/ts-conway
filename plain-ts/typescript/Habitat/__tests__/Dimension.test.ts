import {Dimension} from '@Conway/Habitat/Dimension';

describe('Dimension', () => {

    test('can be created', () => {

        const dimension = new Dimension(2, 3);

        expect(dimension).toBeInstanceOf(Dimension);

    });


    test('getRows() can be retrieved', () => {

        const dimension = new Dimension(2, 3);

        expect(dimension.getRows()).toBe(2);

    });

    test('getCols() can be retrieved', () => {

        const dimension = new Dimension(2, 3);

        expect(dimension.getCols()).toBe(3);

    });

    test('throw execption for getRows() value less then 1', () => {

        expect(() => {
            new Dimension(0, 3);
        }).toThrowErrorMatchingSnapshot();

        expect(() => {
            new Dimension(-1, 3);
        }).toThrowErrorMatchingSnapshot();

    });

    test('throw exception for getCols() value less then 1', () => {

        expect(() => {
            new Dimension(3, 0);
        }).toThrowErrorMatchingSnapshot();

        expect(() => {
            new Dimension(3, -1);
        }).toThrowErrorMatchingSnapshot();

    });

});
