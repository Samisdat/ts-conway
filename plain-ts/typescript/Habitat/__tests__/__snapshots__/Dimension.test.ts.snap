// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dimension throw exception for getCols() value less then 1 1`] = `"cols must be at least 1"`;

exports[`Dimension throw exception for getCols() value less then 1 2`] = `"cols must be at least 1"`;

exports[`Dimension throw execption for getRows() value less then 1 1`] = `"rows must be at least 1"`;

exports[`Dimension throw execption for getRows() value less then 1 2`] = `"rows must be at least 1"`;
